capreolus/extractor/__init__.py

```diff
@@ -194,51 +198,180 @@ def config():
+        return tf/ total_terms
+
+    def id2vec(self, qid, posid, negid=None, query=None):#todo (ask) where is it used?
+        if query is not None:
+            if qid is None:
+                query = self["tokenizer"].tokenize(query)
+            else:
+                raise RuntimeError("received both a qid and query, but only one can be passed")
+        else:
+            query = self.qid2toks[qid]
 
-    def id2vec(self, qid, posid, negid=None, query=None):
+        return {"qid": qid, "posdocid": posid, "negdocid": negid}
+
+
+class DocStatsEmbedding(DocStats):
+    name = "docstatsembedding"
+    dependencies = {# TODO is this okay like this? if this is changed here, would the parent functions also use differently??
+        "index": Dependency(module="index", name="anserini", config_overrides={"indexstops": True, "stemmer": "none"}),
+        # "tokenizer": Dependency(module="tokenizer", name="anserini", config_overrides={"keepstops": False}),
+        "tokenizerquery": Dependency(module="tokenizer", name="spacy", config_overrides={"keepstops": False, 'removesmallerlen': 2}), #removesmallerlen is actually only used for user profile (not the short queries) but I cannot separate them
+        "tokenizer": Dependency(module="tokenizer", name="spacy", config_overrides={"keepstops": False}),
+    }
+
+    embed_names = {
+        "glove6b": "glove-wiki-gigaword-300",
+        "glove6b.50d": "glove-wiki-gigaword-50",
+        "w2vnews" : "word2vec-google-news-300",
+        "fasttext": "fasttext-wiki-news-subwords-300", #TODO: "Wikipedia 2017, UMBC webbase corpus and statmt.org news dataset (16B tokens)" is this model the same as one used with Magnitude?
+
+    }
+
+    def exist(self):
+        return hasattr(self, "similarity_matrix")
+
+    @staticmethod
+    def config():
+        embeddings = "w2vnews"
+
+    def _get_pretrained_emb(self):
+        gensim_cache = CACHE_BASE_PATH / "gensim/"
+        os.environ['GENSIM_DATA_DIR'] = str(gensim_cache.absolute())
+
+        import gensim
+        import gensim.downloader as api
+
+        model_path = api.load(self.embed_names[self.cfg["embeddings"]], return_path=True)
+        return gensim.models.KeyedVectors.load_word2vec_format(model_path, binary=True)
+
+    def create(self, qids, docids, topics, qdocs=None):
+        if self.exist():
+            return
+
+        super().create(qids, docids, topics, qdocs)
+
+        logger.debug("loading embedding")
+        self.emb_model = self._get_pretrained_emb()
+
+
+    def get_term_occurrence_probability(self, qterm, docterm, docid, threshold):
+        nu = self.emb_model.similarity(qterm, docterm) if (self.emb_model.__contains__(qterm) and self.emb_model.__contains__(docterm)) else 0
+        if nu < threshold:
+            return 0
+
+        de = 0
+        for term in self.doc_tf[docid].keys():#TODO could I precalc this? There is the threshold which is the reranker parameter... Should I make it the extractor parameter??
```

Comment:
Yeah, you could move it to be an extractor parameter if that simplifies things. It would be just as easily configurable that way

Reply:
At the moment, I think it's better to keep it as it is. 
Because 1- the similarity threshold (which is a parameter of the LM+Embedding Ranker) kind of should be the parameter of the ranker and not the extractor [hence we can change it at runtime].
And 2- that it is calculated here only when the numerator is greater than the threshold. So we don't need to precalc this for all document terms probably.

Maybe, alternatively, I could store them when each of them is calculated for the first time and then reuse it??

Reply:
It should still be possible to change it at runtime by setting eg `reranker.extractor.similarity_threshold=0.4`, but I think leaving it as a ranker parameter is also fine

Reply:
ah, good to know this :+1 I didn't know that we could change other modules' parameters at runtime.